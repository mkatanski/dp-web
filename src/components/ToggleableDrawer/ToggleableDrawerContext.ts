import React from "react";
import { createContextWithPartial } from "utils/context/createContextWithPartial";

export type ToggleableDrawerContextElements = {
  setDrawerState: React.Dispatch<React.SetStateAction<boolean>>;
  drawerState: boolean;
};

export const ToggleableDrawerContext = createContextWithPartial<
  ToggleableDrawerContextElements
>({});
