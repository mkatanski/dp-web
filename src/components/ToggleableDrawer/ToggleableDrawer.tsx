import React, { useState } from "react";
import { Drawer } from "@material-ui/core";
import { ToggleableDrawerContext } from "./ToggleableDrawerContext";

export type RenderButtonProps = {
  setDrawerState: React.Dispatch<React.SetStateAction<boolean>>;
};

export type ToggleableDrawerProps = React.PropsWithChildren<{
  drawerAnchor?: "left" | "top" | "right" | "bottom";
  renderButton: (props: RenderButtonProps) => React.ReactNode;
}>;

type DrawerOnClose = (
  event: {},
  reason: "backdropClick" | "escapeKeyDown"
) => void;

export const ToggleableDrawer: React.FC<ToggleableDrawerProps> = ({
  children,
  drawerAnchor,
  renderButton
}: ToggleableDrawerProps) => {
  const [isOpen, setDrawerState] = useState(false);

  const handleCloseDrawer: DrawerOnClose = () => {
    setDrawerState(false);
  };

  return (
    <ToggleableDrawerContext.Provider
      value={{ setDrawerState, drawerState: isOpen }}
    >
      {renderButton({ setDrawerState })}
      <Drawer
        data-testid="toggleable-drawer"
        anchor={drawerAnchor}
        open={isOpen}
        onClose={handleCloseDrawer}
      >
        {children}
      </Drawer>
    </ToggleableDrawerContext.Provider>
  );
};
