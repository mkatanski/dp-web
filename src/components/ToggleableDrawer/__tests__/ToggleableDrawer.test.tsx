/* eslint-disable react/display-name */
import React, { useContext, useEffect } from "react";
import enzyme from "enzyme";
import TestRenderer from "react-test-renderer";

import { ToggleableDrawer } from "../ToggleableDrawer";
import { ToggleableDrawerContext } from "../ToggleableDrawerContext";

describe("components", () => {
  describe("ToggleableDrawer", () => {
    it("should render ToggleableDrawerView component with children", () => {
      const tree = TestRenderer.create(
        <ToggleableDrawer renderButton={() => <button />}>
          child
        </ToggleableDrawer>
      );
      expect(tree).toMatchSnapshot();
    });

    it("should open drawer on button click", () => {
      const component = enzyme.mount(
        <ToggleableDrawer
          drawerAnchor="bottom"
          renderButton={({ setDrawerState }) => (
            <button onClick={() => setDrawerState(true)} />
          )}
        >
          child
        </ToggleableDrawer>
      );

      component.find("button").simulate("click");
      expect(component).toMatchSnapshot();
    });

    it("should open drawer via context", () => {
      const InternalComponent = () => {
        const { setDrawerState } = useContext(ToggleableDrawerContext);

        useEffect(() => {
          setDrawerState(true);
        }, [setDrawerState]);

        return <p>Internal Component</p>;
      };

      const component = enzyme.mount(
        <ToggleableDrawer
          drawerAnchor="bottom"
          renderButton={({ setDrawerState }) => (
            <button onClick={() => setDrawerState(true)} />
          )}
        >
          <InternalComponent />
        </ToggleableDrawer>
      );

      expect(component).toMatchSnapshot();
    });
  });
});
