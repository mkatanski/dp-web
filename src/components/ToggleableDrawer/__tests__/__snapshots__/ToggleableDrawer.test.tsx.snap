// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components ToggleableDrawer should open drawer on button click 1`] = `
<ToggleableDrawer
  drawerAnchor="bottom"
  renderButton={[Function]}
>
  <button
    onClick={[Function]}
  />
  <Drawer
    anchor="bottom"
    data-testid="toggleable-drawer"
    onClose={[Function]}
    open={true}
  >
    <div
      data-props={
        Object {
          "anchor": "bottom",
          "data-testid": "toggleable-drawer",
          "onClose": [Function],
          "open": true,
        }
      }
      id="MUI.Drawer"
    >
      child
    </div>
  </Drawer>
</ToggleableDrawer>
`;

exports[`components ToggleableDrawer should open drawer via context 1`] = `
<ToggleableDrawer
  drawerAnchor="bottom"
  renderButton={[Function]}
>
  <button
    onClick={[Function]}
  />
  <Drawer
    anchor="bottom"
    data-testid="toggleable-drawer"
    onClose={[Function]}
    open={true}
  >
    <div
      data-props={
        Object {
          "anchor": "bottom",
          "data-testid": "toggleable-drawer",
          "onClose": [Function],
          "open": true,
        }
      }
      id="MUI.Drawer"
    >
      <InternalComponent>
        <p>
          Internal Component
        </p>
      </InternalComponent>
    </div>
  </Drawer>
</ToggleableDrawer>
`;

exports[`components ToggleableDrawer should render ToggleableDrawerView component with children 1`] = `
Array [
  <button />,
  <div
    data-props={
      Object {
        "anchor": undefined,
        "data-testid": "toggleable-drawer",
        "onClose": [Function],
        "open": false,
      }
    }
    id="MUI.Drawer"
  >
    child
  </div>,
]
`;
